pyi.py

```diff
@@ -442,6 +500,52 @@ def _visit_function(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> None:
                     continue
             self.error(statement, Y010)
 
+        # Rest of this method is devoted to finding TypeVars that should be _typeshed.Self, but aren't
+
+        if self.in_class:
+            self.check_self_typevars(node)
+
+    def check_self_typevars(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> None:
+        pos_or_keyword_args = node.args.args
+
+        if not pos_or_keyword_args:
+            return
+        return_annotation = node.returns
+
+        if not isinstance(return_annotation, ast.Name):
+            return
+        first_arg_annotation = pos_or_keyword_args[0].annotation
+
+        if not isinstance(first_arg_annotation, (ast.Name, ast.Subscript)):
+            return
+
+        if node.name == "__new__":
+            self._check_class_method_for_bad_typevars(
+                method=node,
+                first_arg_annotation=first_arg_annotation,
+                return_annotation=return_annotation,
+            )
+            return
+
+        for decorator in node.decorator_list:
+            if isinstance(decorator, ast.Name):
+                decorator_name = decorator.id
+                if decorator_name == "classmethod":
+                    self._check_class_method_for_bad_typevars(
+                        method=node,
+                        first_arg_annotation=first_arg_annotation,
+                        return_annotation=return_annotation,
+                    )
+                    return
+                elif decorator_name == "staticmethod":
+                    return
```

Comment:
A list like `decorator_names = [decor.id for decor in node.decorator_list if isinstance(decor, ast.Name)]` would make this code a lot more readable.